src/OrchardCore/OrchardCore.ReCaptcha.Core/TagHelpers/ReCaptchaTagHelper.cs

```diff
@@ -1,42 +1,22 @@
-using System.Globalization;
-using System.Linq;
 using System.Threading.Tasks;
-using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.TagHelpers;
 using Microsoft.AspNetCore.Razor.TagHelpers;
-using Microsoft.Extensions.DependencyInjection;
-using Microsoft.Extensions.Localization;
-using Microsoft.Extensions.Logging;
-using Microsoft.Extensions.Options;
-using OrchardCore.Localization;
-using OrchardCore.Modules;
 using OrchardCore.ReCaptcha.ActionFilters;
-using OrchardCore.ReCaptcha.ActionFilters.Detection;
-using OrchardCore.ReCaptcha.Configuration;
-using OrchardCore.ResourceManagement;
+using OrchardCore.ReCaptcha.Services;
 
 namespace OrchardCore.ReCaptcha.TagHelpers
 {
     [HtmlTargetElement("captcha", TagStructure = TagStructure.WithoutEndTag)]
     [HtmlTargetElement("captcha", Attributes = "mode,language", TagStructure = TagStructure.WithoutEndTag)]
     public class ReCaptchaTagHelper : TagHelper
     {
-        private readonly IResourceManager _resourceManager;
-        private readonly IHttpContextAccessor _httpContextAccessor;
-        private readonly ReCaptchaSettings _settings;
-        private readonly ILogger _logger;
-        private readonly ILocalizationService _localizationService;
-        private readonly IStringLocalizer S;
+        private readonly ReCaptchaService _reCaptchaService;
 
-        public ReCaptchaTagHelper(IOptions<ReCaptchaSettings> optionsAccessor, IResourceManager resourceManager, ILocalizationService localizationService, IHttpContextAccessor httpContextAccessor, ILogger<ReCaptchaTagHelper> logger, IStringLocalizer<ReCaptchaTagHelper> localizer)
+        public ReCaptchaTagHelper(ReCaptchaService reCaptchaService)
```

Comment:
Use a lazy service resolution since otherwise you'll keep a long-lived reference to this service, that has references to transient services.

`IHttpContextAccessor`

Reply:
So you want the same way of injecting `IServiceProvider` and getting `ReCaptchaService` from it?
I don't understand what is the benefit. Anyway, the services are transient. Can something change in the same request?
How the tag helper can be not transient?

Reply:
@sebastienros @xperiandri

Tag helpers are registered and resolved as transient services only if we call `.AddTagHelpersAsServices()`, otherwise, as in our case, mvc uses the `ActivatorUtilities` to instantiate a type with constructor arguments provided directly and / or from a service provider that here is the current `RequestServices`.
https://github.com/dotnet/aspnetcore/blob/a08aa2dd318956d5b7295561d9423252f6071a52/src/Mvc/Mvc.Razor/src/Infrastructure/DefaultTagHelperActivator.cs#L25

So, even they are instanciated, not resolved from the DI, they can be considered as transient services, the only difference is if a tag helper is itself an `IDisposable` (not the case here) it would not be disposed when the service provider will be disposed, but the services defined in its ctor will be still disposed as they have been resolved from the DI, here through `RequestServices`.

One thing we have to take care is when a transient service injects a scoped service (or indirectly through another transient) and may be also injected by a singleton (it is detected at runtime), but this is not the case here, and we have many tag helpers that inject scoped services.

**So here, and in any of our tag helpers, we can inject any service of any lifetime.** I think that @sebastienros was thinking about the related liquid tag implementation where the services need to be lazily resolved.

Note: We register our tag helpers (not the mvc ones) as transient services with our `AddTagHelpers()`, but only to resolve them from the DI while under liquid, not under razor where they are still instanciated on demand.

Note: In fact we are running in an isolated shell scope with its own service provider, but we made `RequestServices` aware of the current shell scope, so that it returns the right service provider.

Reply:
@jtkech, so that we don't need to call?
https://github.com/OrchardCMS/OrchardCore/blob/c803cc0a23d1da875f554a029505d4da997a8cea/src/OrchardCore/OrchardCore.ReCaptcha.Core/ServiceCollectionExtensions.cs#L27

Reply:
ok @jtkech it was me arguing that we shouldn't inject transient into tag helpers, because I was under the impression that asp.net core sometimes cached them during requests.

If you are certain, fine with me. (the concern was the service wraps an httpclient factory, which should be allowed to manage itself transiently)

@Diman96 Keep the `services.AddTagHelpers()` please, as it is already in use for users, through the liquid taghelper bridge.

Reply:
@xperiandri @deanmarcussen 

For a given tag helper type, this is the factory that is statically cached, but not its instantiation.

Just tried by injecting a transient in our script tag helper, tested under razor and liquid, in both cases and in the same request scope, a new instance of the tag helper is always created (as a transient service), each one injects the same ressource manager instance that is a scoped service, but injects a different instance of the transient service.

![Transients](https://user-images.githubusercontent.com/8586360/106341098-570fb700-629c-11eb-95d7-a1cfe91d0b3d.png)